### PR TITLE
Feature/calendar to kg using kafka

### PR DESCRIPTION
**Pull request by mshash8**

### CODE REVIEW

File: api/main.py [hunk 1]
```diff
@@ -59,6 +66,54 @@ def require(keys: list[str], type = str) -> str:
     keys = "' or '".join(keys)
     flask.abort(400, f"Missing required parameter '{keys}'")
 
+@app.route('/calendar', methods=['GET'])
+def calendar() -> str:
```
Comment (knordstrom) [inline api/main.py:70]: Are these commented sections relevant to the method?

File: processor/main.py [hunk 2]
```diff
@@ -36,14 +37,110 @@ def write_to_vdb(mapped: list) -> None:
...
+                status = ""
+                start = ""
+                end = ""
+                event_summary = ""
+                event_description = ""
+                event_recurring_id = ""
+                # get event information (meeting information)
+                event_id = event["id"]
+                if "status" in event:
+                    status = event["status"]
```
Comment (knordstrom) [inline processor/main.py]: You can do event.get("status") to avoid the dereference error

File: processor/main.py [hunk 3]
```diff
@@ -36,14 +37,110 @@ def write_to_vdb(mapped: list) -> None:
...
+                    event_recurring_id = event["recurringEventId"]
+                if "start" in event:
+                    start = event["start"].get("dateTime", event["start"].get("date"))
+                if "end" in event:
+                    end = event["end"].get("dateTime", event["end"].get("date"))
+                if "summary" in event:
+                    event_summary = event["summary"]
+                if "description" in event:
+                    event_description = event["description"]
+                event_dict = {"id": event_id, "start": start, "end": end, "description": event_description,
```
Comment (knordstrom) [inline processor/main.py]: I would collapse lines 60 through 80 into a single line that leverages event.get(..) to create the dict and avoid any intermediate variables unless you explicitly need them. Short methods (<20-30 lines) are preferable for readability purposes

File: processor/main.py [hunk 4]
```diff
@@ -36,14 +37,110 @@ def write_to_vdb(mapped: list) -> None:
             if w is not None:
                 w.close()
 
+def write_to_neo4j(events) -> None:
```
Comment (knordstrom) [inline processor/main.py]: I think this function here would be better off in a class elsewhere where it can be reused or made into a library etc.

It's also doing multiple things, ie. It connects to the database, it constructs the data object, and then it writes to the db. Try to encapsulate each of those separate actions in their own logical constructs so they can be composed here.

One mental mnemonic I use is "Do I feel compelled to write a comment describing what the next lines are doing?" If yes, it's a separate method the name of which should describe what I would have put in that comment. 

I think it's counterintuitive and maybe even contrary to some of what gets taught, but the best code has only docstrings, never online comments.

File: docker/docker-compose.yml [hunk 5]
```diff
@@ -71,8 +71,17 @@ services:
   processor:
     image: context-processor:0.1.1
     environment:
-      KAFKA_BROKER: "kafka1:29092"
+      KAFKA_BROKER: "host.docker.internal:29092"
```
Comment (knordstrom) [inline docker/docker-compose.yml:74]: It is weird this is required on yours and not mine. The interface really should be kafka1, at least according to my experience; I guess we'll find out :)

**Review by knordstrom — CHANGES_REQUESTED** [review]

File: processor/main.py (hunk 4 above)
Comment (mshash8) [reply processor/main.py]: @knordstrom I agree with this!
Where do you think I could create a Class for this? I also wanted to create a Class for each node type like Person and Event. What do you think about this?

File: docker/docker-compose.yml (hunk 5 above)
Comment (mshash8) [reply docker/docker-compose.yml:74]: I'll try it as kafka1. I changed this when I was trying to debug an issue. It is also working with kafka1 so I'll update this.

File: api/main.py (hunk 1 above)
Comment (mshash8) [reply api/main.py:70]: The comments are just to explain the authorization procedure. I can remove them if it is adding too much fluff.

File: api/main.py [hunk 6]
```diff
@@ -59,6 +66,54 @@ def require(keys: list[str], type = str) -> str:
...
+    count = flask.request.args.get('n', None, int)
+    SCOPES = ["https://www.googleapis.com/auth/calendar.readonly"]
+    creds = None
+    if os.path.exists("token.json"):
+        creds = Credentials.from_authorized_user_file("token.json", SCOPES)
+    # If there are no (valid) credentials available, let the user log in.
+    if not creds or not creds.valid:
```
Comment (knordstrom) [inline api/main.py:77]: This segment in here is almost identical to the __authenticate code in the gmail object in the library. Let's leave it for now, I'll refactor to combine. I actually like the use of json better than the pickle code that I have already.

Comment (knordstrom) [reply api/main.py:77]: Generally speaking, it's a good idea to 1. keep all methods short (at 40 lines this on is getting long) and 2. keep logic out of the API/front end layer.

Notice how you made inline comments above? Those should be triggers that you really want the code to live somewhere more modular and targeted at a single behavior.

File: library/neo4j.py [hunk 7]
```diff
@@ -25,10 +38,87 @@ def get_schedule(self, email: str, start_time: datetime, end_time: datetime) ->
...
+        for record in events:
+            event = record.value
+            event_dict = self.extract_event_info(event)
+            events_list.append(event_dict)
+            self.extract_attendees_info(event, event_dict, person_list, attendance_list)
+        self.add_to_db(person_list, events_list, attendance_list)
+
+    def extract_event_info(self, event):
+        event_dict = {
+            "id": event.get("id"),
```
Comment (knordstrom) [inline library/neo4j.py:72]: Nice, I like the changes. These are nicely readable methods.

**Review by knordstrom — APPROVED** [review]
LGTM
